### PR TITLE
Auto choose conv implementation

### PR DESCRIPTION
**Pull request by tpoisonooo**
"Engineering is magic."

1. add auto pack model to choose the fastest convolution implementation automatically;
2. add relative design doc;
3. remove ncnn2mem/convolution compile warnings;
4. add convolution1x1.h namespace.

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1085) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1085) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: tpoisonooo<br/>:x: lichengcheng<br/><hr/>**lichengcheng** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1085) it.</sub>

**Comment by tpoisonooo** [comment]
Repeat`example/squeezenet` 50 times on RK3399.

> user@X3399:~/GitProjects/ncnn/build/examples$ time ./squeezenet ~/a.jpg
638 = 0.044468
869 = 0.040067
837 = 0.039842
real    0m8.731s
user    0m7.330s
sys 0m1.393s

After auto pack model:
>638 = 0.044468
869 = 0.040067
837 = 0.039842
real    0m8.449s
user    0m7.107s
sys 0m1.340s

This method save `6ms` on squeezenet_v1.1 .

File: tools/ncnnoptimize.cpp [hunk 1]
```diff
@@ -69,6 +70,29 @@
 #include "layer/yolodetectionoutput.h"
 #include "layer/yolov3detectionoutput.h"
 
+#if (defined __aarch64__) && (defined ANDROID)
```
Comment (Howave) [inline tools/ncnnoptimize.cpp]: it breaks the compilation for armv7 and other os

Comment (tpoisonooo) [reply tools/ncnnoptimize.cpp]: nooop, It has been compiled on Raspberry Pi.

File: src/layer/arm/convolution_arm.cpp [hunk 2]
```diff
@@ -84,9 +85,12 @@ int Convolution_arm::create_pipeline(const Option& opt)
         activation->create_pipeline(opt_cpu);
     }
 
+
```
Comment (nihui) [inline src/layer/arm/convolution_arm.cpp]: empty lines changes

File: src/layer/arm/convolution_arm.cpp [hunk 3]
```diff
@@ -84,9 +85,12 @@ int Convolution_arm::create_pipeline(const Option& opt)
...
+
     use_winograd3x3 = false;
     use_sgemm1x1 = false;
 
+
```
Comment (nihui) [inline src/layer/arm/convolution_arm.cpp]: empty lines changes

File: src/layer/arm/convolution_arm.cpp [hunk 4]
```diff
@@ -137,6 +141,42 @@ int Convolution_arm::create_pipeline(const Option& opt)
...
+                break;
+            case 4:
+                // direct
+                break;
+            case 5:
+                // conv3x3s2
+                num_input = weight_data_size / 9 / num_output;
+                conv3x3s2_transform_kernel_neon(weight_data, weight_3x3s2_data, num_input, num_output);
+            default:
+                break;
```
Comment (nihui) [inline src/layer/arm/convolution_arm.cpp]: return -1 to indicate invalid impl_type value error

File: src/layer/arm/convolution_arm.cpp [hunk 5]
```diff
@@ -596,28 +636,57 @@ int Convolution_arm::forward(const Mat& bottom_blob, Mat& top_blob, const Option
...
+                conv_im2col_sgemm_neon(bottom_blob_bordered, top_blob, weight_sgemm_data, bias_data, kernel_w, kernel_h, stride_w, stride_h, opt);
+                break;
+            case 4:
+                conv(bottom_blob_bordered, top_blob, weight_data, bias_data, opt);
+                break;
+            case 5:
+                conv3x3s2_packed_neon(bottom_blob_bordered, top_blob, weight_3x3s2_data, bias_data, opt);
+                break;
+            default:
+                assert(0);
```
Comment (nihui) [inline src/layer/arm/convolution_arm.cpp]: assert(0) to terminate program in library code is not acceptable, return error code and let the caller to handle it

File: src/layer/convolution.h [hunk 6]
```diff
@@ -74,6 +74,9 @@ class Convolution : public Layer
     // merge de/requantize op into convolution op
     std::vector<float> dequantize_scales;
     std::vector<float> requantize_scales;    
+
+    // implementation type, -1 means do not use auto pack model
```
Comment (nihui) [inline src/layer/convolution.h]: the default value is 0, not -1

File: tools/ncnnoptimize.cpp [hunk 7]
```diff
@@ -69,6 +70,29 @@
...
 
+#if defined(__aarch64__) && defined(LINUX)
+namespace ncnn{
+#include "layer/arm/convolution_1x1.h"
+#include "layer/arm/convolution_2x2.h"
+#include "layer/arm/convolution_3x3.h"
+#include "layer/arm/convolution_4x4.h"
+#include "layer/arm/convolution_5x5.h"
+#include "layer/arm/convolution_7x7.h"
+#include "layer/arm/convolution_sgemm.h"
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: internal kernel functions are not used anymore, these includes are useless  :)

File: tools/ncnnoptimize.cpp [hunk 8]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
     int fwrite_weight_data(const ncnn::Mat& data, FILE* bp);
 
     int save(const char* parampath, const char* binpath);
+
+#if defined(__aarch64__) && defined(LINUX)
+    void gauss_random(ncnn::Mat &m);
+    void find_fastest_fp32_conv(const char* name, int w, int h, int c);
+    int support_fp32_conv_type(ncnn::Convolution* op, ncnn::Mat& mat, int type);
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: const++

```
int support_fp32_conv_type(ncnn::Convolution* op, ncnn::Mat& mat, int type);
```

```
int support_fp32_conv_type(const ncnn::Convolution* op, const ncnn::Mat& mat, int type) const;
```

File: tools/ncnnoptimize.cpp [hunk 9]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
 };
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]:
```
#endif // defined(__aarch64__) && defined(LINUX)

};

#if defined(__aarch64__) && defined(LINUX)
```

File: tools/ncnnoptimize.cpp [hunk 10]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
 
+void NetOptimize::gauss_random(ncnn::Mat &m)
+{
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::normal_distribution<float> d(1.0f, 1.0f);
+    int size = m.c * m.h * m.w;
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: channels may have gaps
access total element count via m.total()

File: tools/ncnnoptimize.cpp [hunk 11]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
+    for (int i = 0; i < size; ++i)
+    {
+        (static_cast<float*>(m.data))[i] = d(gen);
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: simple fp32 assignment candy
```
m[i] = d(gen);
```

File: tools/ncnnoptimize.cpp [hunk 12]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
+            for (int type = 1; type <= 5; ++type)
+            {
+                int support = support_fp32_conv_type(op, bottom_blob, type);
+                if (support < 1)
+                {
+                    // implementation type mismatch convolution configuration, skip
+                    continue;
+                }
+
+                op->impl_type = type;
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: this line is useless  :]

File: tools/ncnnoptimize.cpp [hunk 13]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
+
+                auto start = std::chrono::high_resolution_clock::now();
+                const int NREPEATS = 20;
+                op->create_pipeline(opt);
+                for (int repeat = 0; repeat < NREPEATS; ++repeat)
+                {
+                    op->forward(top_blob, bottom_blob, opt);
+                }
+
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: though it seems to work properly
it is better calling op->destroy_pipeline(opt); after using this operator

File: tools/ncnnoptimize.cpp [hunk 14]
```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
...
+    // kernel_size x stride
+    int kernel = op->kernel_h,
+        stride = op->stride_h;
+    // if match prequisation
+    switch(type)
+    {
+        case 1:
+            // winograd
+            if (kernel != 3 || stride != 1){
+                return -1.0;
```
Comment (nihui) [inline tools/ncnnoptimize.cpp]: return -1

and also the following code blocks

File: tools/ncnnoptimize.cpp (hunk 12 above)
Comment (tpoisonooo) [reply tools/ncnnoptimize.cpp]: This line setup convolution type for `op->forward`.

**Pull request merged**